shared/constants/metametrics.js

```diff
@@ -0,0 +1,213 @@
+// Type Imports
+/**
+ * @typedef {import('../../app/scripts/lib/enums').EnvironmentType} EnvironmentType
+ */
+
+// Type Declarations
+/**
+ * Used to attach context of where the user was at in the application when the
+ * event was triggered. Also included as full details of the current page in
+ * page events.
+ * @typedef {Object} MetaMetricsPageObject
+ * @property {string} [path] - the path of the current page (e.g /home)
+ * @property {string} [title] - the title of the current page (e.g 'home')
+ * @property {string} [url] - the fully qualified url of the current page
+ */
+
+/**
+ * For metamask, this is the dapp that triggered an interaction
+ * @typedef {Object} MetaMetricsReferrerObject
+ * @property {string} [url] - the origin of the dapp issuing the
+ *  notification
+ */
+
+/**
+ * We attach context to every meta metrics event that help to qualify our
+ * analytics. This type has all optional values because it represents a
+ * returned object from a method call. Ideally app and userAgent are
+ * defined on every event. This is confirmed in the getTrackMetaMetricsEvent
+ * function, but still provides the consumer a way to override these values if
+ * necessary.
+ * @typedef {Object} MetaMetricsContext
+ * @property {Object} app
+ * @property {string} app.name - the name of the application tracking the event
+ * @property {string} app.version - the version of the application
+ * @property {string} userAgent - the useragent string of the user
+ * @property {MetaMetricsPageObject} [page] - an object representing details of
+ *  the current page
+ * @property {MetaMetricsReferrerObject} [referrer] - for metamask, this is the
+ *  dapp that triggered an interaction
+ */
+
+/**
+ * @typedef {Object} MetaMetricsEventPayload
```

Comment:
Does it still make sense for these type definitions to live here, rather than alongside the function they're used with? I'm not sure I see the benefit - I found myself switching back and forth a lot to try and match things up.

Reply:
I use these types in the frontend as well to type the methods that are used for tracking the events through the background, so at least to me, it makes sense to store them in a shared location. Though I guess there is an argument to be made for "storing them in shared/ means you always have to flip files to look them up, instead of importing them directly from the MetaMetrics controller file where you'd only have to look them up from the frontend".

I honestly flip-flopped on this the entire time I was writing them.

Reply:
Ah I see - I mistakenly thought this type was only used in the background. I see it's used in the context as well. Shared makes perfect sense then.